test/common/network/lc_trie_test.cc

```diff
@@ -0,0 +1,105 @@
+#include "common/common/utility.h"
+#include "common/network/address_impl.h"
+#include "common/network/cidr_range.h"
+#include "common/network/lc_trie.h"
+
+#include "gtest/gtest.h"
+
+namespace Envoy {
+namespace Network {
+namespace LcTrie {
+typedef std::pair<std::string, std::vector<Address::CidrRange>> IpTagData;
+
+namespace {
+
+std::vector<IpTagData>
+generateTestCase(const std::vector<std::vector<std::string>>& cidr_range_strings) {
+  std::vector<std::pair<std::string, std::vector<Address::CidrRange>>> output;
+  IpTagData ip_tags;
+  for (size_t i = 0; i < cidr_range_strings.size(); i++) {
+    ip_tags.first = fmt::format("tag_{0}", i);
+    for (size_t j = 0; j < cidr_range_strings[i].size(); j++) {
+      ip_tags.second.push_back(Address::CidrRange::create(cidr_range_strings[i][j]));
+    }
+    output.push_back(ip_tags);
+    ip_tags.second.clear();
+  }
+  return output;
+};
+} // namespace
+
+// TODO(ccaraman): Add a performance benchmark test.
+
+TEST(LcTrie, Ipv4) {
+
+  std::vector<std::vector<std::string>> cidr_range_strings = {
+      {"1.2.3.4/24", "10.255.255.255/32"},
+      {"54.233.128.0/17", "205.251.192.100/26", "52.220.191.10/30"}};
+
+  LcTrie trie(generateTestCase(cidr_range_strings));
+
+  std::vector<std::pair<std::string, std::string>> test_case = {
+      {"205.251.192.100", "tag_1"},
+      {"18.232.0.255", ""},
+      {"10.255.255.255", "tag_0"},
+      {"52.220.191.10", "tag_1"},
+  };
+
+  for (const auto& kv : test_case) {
+    EXPECT_EQ(kv.second, trie.search(Utility::parseInternetAddress(kv.first)));
+  }
+
+  // Check Ipv6 path doesn't crash when the ipv6 trie is empty.
+  EXPECT_EQ("", trie.search(Utility::parseInternetAddress("::1")));
+}
+
+TEST(LcTrie, Ipv6) {
+  std::vector<std::vector<std::string>> cidr_range_strings = {
+      {"2406:da00:2000::/40", "::1/128"}, {"2001:abcd:ef01:2345:6789:abcd:ef01:234/64"}};
+  LcTrie trie(generateTestCase(cidr_range_strings));
+
+  std::vector<std::pair<std::string, std::string>> test_case = {{"2400:ffff:ff00::", ""},
+                                                                {"2406:da00:2000::1", "tag_0"},
+                                                                {"2001:abcd:ef01:2345::1", "tag_1"},
+                                                                {"::1", "tag_0"}};
+
+  for (const auto& kv : test_case) {
+    EXPECT_EQ(kv.second, trie.search(Utility::parseInternetAddress(kv.first)));
+  }
+
+  // Check the Ipv4 path doesn't crash when the Ipv4 trie is empty.
+  EXPECT_EQ("", trie.search(Utility::parseInternetAddress("1.2.3.4")));
+}
+
+TEST(LcTrie, BothIpvVersions) {
```

Comment:
In practice, will IPv4 and IPv6 need separate trees? I'm thinking about the case where a user has configured an IPv4 prefix with length 24, and then they get a request from an IPv6 address whose first 24 bits happen to match the prefix. In that scenario, the user will probably want the IPv6 address to _not_ match the IPv4 prefix.

Reply:
Oops, never mind; I just realized the LCTrie maintains two separate LCTrieInternal objects, so it won't accidentally match v6 addresses against v4 prefixes.

Reply:
I'll improve the comments for the LcTrie to make this clearer. Thank you for the feedback.